### PR TITLE
Submit Analytics Data (EXPOSUREAPP-4823)

### PR DESCRIPTION
**Pull request by kolyaopahle**
### Description
This PR contains the backend logic for Analytics submission
It should only be merged after all other Analytics PRs are merged as it will be used to combine front and backend interactions.

### Testing
There are quite a few options in the Data Donation tab of the Tester Menu to already test functionality, but there is currently no working testing environment providing the necessary backend.
You could use the cwa-app-cli project which has support for the backend or check logcat for errors

### CODE REVIEW

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.File
+import javax.inject.Inject
+
+class DefaultLastAnalyticsSubmissionLogger @Inject constructor(
+    @AppContext private val context: Context,
+    private val dispatcherProvider: DispatcherProvider,
+    @BaseGson private val baseGson: Gson
+) : LastAnalyticsSubmissionLogger {
+    private val analyticsDir = File(context.cacheDir, "analytics_storage")
+    private val analyticsFile = File(analyticsDir, "analytics.bin")
```
Comment (d4rken) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:31]: Minor, `analytics.bin` seems very generic, isn't it `last_analytics.bin`?

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            try {
+                gson.toJson(dataObject, analyticsFile)
+            } catch (e: Exception) {
+                Timber.e(e, "Failed to store analytics data.")
+            }
+        }
+
+    override suspend fun getLastAnalyticsData(): LastAnalyticsSubmission? = withContext(dispatcherProvider.IO) {
+        try {
+            gson.fromJson(analyticsFile)
```
Comment (d4rken) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:60]: Based on issues we had with the exposure detection which also stores json to file:
We should prevent empty files from causing invalid data classes being created by GSON (examples in `ExposureDetectionTrackerStorageTest`).

Despite not being allowed GSON can create data classes with null fields that shouldn't be null (unsafe init), which the JVM will only notice once accessed.

TL;DR:  `.also { requireNotNull(it.timestamp) }`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt [hunk 3]
```diff
@@ -4,20 +4,59 @@ import dagger.Reusable
...
 @Reusable
 class AnalyticsConfigMapper @Inject constructor() : AnalyticsConfig.Mapper {
     override fun map(rawConfig: AppConfigAndroid.ApplicationConfigurationAndroid): AnalyticsConfig {
-        // TODO
-        return AnalyticsConfigContainer()
+        if (!rawConfig.hasPrivacyPreservingAnalyticsParameters() &&
+            !rawConfig.privacyPreservingAnalyticsParameters.hasCommon() &&
+            !rawConfig.privacyPreservingAnalyticsParameters.hasPpac()
+        ) {
+            throw ApplicationConfigurationInvalidException(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:18]: Is the expected behavior to throw here (and discard the whole config), or would it be okay to return a default of "analytics disabled", as analytics being disabled would still allow the rest of the app to function?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt [hunk 4]
```diff
@@ -4,20 +4,59 @@ import dagger.Reusable
...
+                    .probabilityToSubmitExposureWindows
+            )
+        }
+
     data class AnalyticsConfigContainer(
         override val safetyNetRequirements: SafetyNetRequirements = SafetyNetRequirementsContainer(),
-        override val probabilityToSubmit: Float = 1f,
-        override val probabilityToSubmitAfterRiskCalculation: Float = 1f,
-        override val probabilityToSubmitNewExposureWindows: Float = 1f
+        override val probabilityToSubmit: Double = 1.0,
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:57]: Should the defaults be 1.0, or would it be better if the defaults were 0.0, i.e. disabled? Does the spec say?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 5]
```diff
@@ -1,34 +1,156 @@
...
+    private val logger: LastAnalyticsSubmissionLogger
 ) {
+    private suspend fun trySubmission(ppaData: PpaData.PPADataAndroid): Boolean {
+        try {
+            val ppaAttestationRequest = PPADeviceAttestationRequest(
+                ppaData = ppaData
+            )
 
-    val isEnabled: Boolean = true
+            val attestation = deviceAttestation.attest(ppaAttestationRequest)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:36]: This only means that SafetyNet passed "generally". The analytics config contains specific requirements that you need to check, i.e.

`attestation.requirePass(analyticsConfig.requirements)`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 6]
```diff
@@ -1,34 +1,156 @@
...
 
+            val ppaContainer = PpaDataRequestAndroid.PPADataRequestAndroid.newBuilder()
+                .setAuthentication(attestation.accessControlProtoBuf)
+                .setPayload(ppaData)
+                .build()
+
+            dataDonationAnalyticsServer.uploadAnalyticsData(ppaContainer)
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:43]: I'd :heart: a log statement before attestation, after attestation pass and after upload.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 7]
```diff
@@ -1,34 +1,156 @@
...
+        val success = trySubmission(analyticsProto)
+
+        contributions.forEach {
+            Timber.d("Finishing contribution: %s", it::class.simpleName)
+            it.finishDonation(success)
+        }
+
+        if (success) {
+            settings.lastSubmittedTimestamp.update {
+                Instant.now()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:87]: `Timestamper` would like to be part of the analytics gang.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 8]
```diff
@@ -1,34 +1,156 @@
...
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToNoUserConsent(): Boolean {
+        return !settings.analyticsEnabled.value
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    suspend fun stopDueToProbabilityToSubmit(): Boolean {
+        val submitRoll = Random.nextDouble(0.0, 1.0)
+        return submitRoll > appConfigProvider.getAppConfig().analytics.probabilityToSubmit
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:104]: `>=`?

What if we have chance 1.0 and roll a 1.0?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 9]
```diff
@@ -1,34 +1,156 @@
...
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToLastSubmittedTimestamp(): Boolean {
+        val lastSubmit = settings.lastSubmittedTimestamp.value ?: return false
+        return lastSubmit.plus(Hours.hours(LAST_SUBMISSION_MIN_AGE_HOURS).toStandardDuration()).isAfterNow
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:110]: `isAfterNow` internally calls `Instant.now` right? `TimeStamper` so we can fake this in tests?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt [hunk 10]
```diff
@@ -1,34 +1,156 @@
...
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToTimeSinceOnboarding(): Boolean {
+        val onboarding = LocalData.onboardingCompletedTimestamp()?.let { Instant.ofEpochMilli(it) } ?: return true
+        return onboarding.plus(Hours.hours(ONBOARDING_DELAY_HOURS).toStandardDuration()).isAfterNow
+    }
+
+    suspend fun submitIfWanted() {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:119]: I guess this is the entry point to the whole thing, right?
Can we sleep better if we guard this with a `Mutex` or are we absolutely sure there will be no overlaps? If yes why are we so sure?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt [hunk 11]
```diff
@@ -5,26 +5,61 @@ import dagger.Provides
...
+import javax.inject.Singleton
 
 @Module
 class AnalyticsModule {
 
-    @IntoSet
+    @Singleton
     @Provides
-    fun newExposureWindows(module: NewExposureWindowsDonor): DonorModule = module
+    fun provideAnalyticsSubmissionApi(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt:25]: Please align with @chris-cwa regarding the http client for the datadonation server, there are some differences in your PRs.

https://github.com/corona-warn-app/cwa-app-android/pull/2311#discussion_r573655003

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt [hunk 12]
```diff
@@ -5,26 +5,61 @@ import dagger.Provides
...
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.converter.protobuf.ProtoConverterFactory
+import javax.inject.Singleton
 
 @Module
 class AnalyticsModule {
 
-    @IntoSet
+    @Singleton
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt:23]: `Singleton` or `Reusable`, or neither?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+@Singleton
+class DataDonationAnalyticsServer @Inject constructor(
+    private val api: Lazy<DataDonationAnalyticsApiV1>
+) {
+
+    suspend fun uploadAnalyticsData(ppaDataRequestAndroid: PpaDataRequestAndroid.PPADataRequestAndroid) {
+        val response = api.get().submitAndroidAnalytics(ppaDataRequestAndroid)
+
+        if (response.code() == 204) {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt:17]: Could I have some logging for each response type/code :relaxed: ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+import de.rki.coronawarnapp.util.worker.InjectedWorkerFactory
+import de.rki.coronawarnapp.worker.BackgroundConstants
+import timber.log.Timber
+
+/**
+ * One time background Analytics worker
+ *
+ * @see DataDonationAnalyticsScheduler
+ */
+class DataDonationAnalyticsOneTimeWorker @AssistedInject constructor(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt:19]: Do we need a onetime worker? Can win with just the periodic worker?

Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt:19]: Why do they both look the same?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+ */
+@Reusable
+class DataDonationAnalyticsScheduler @Inject constructor(
+    private val workManager: WorkManager,
+    private val workBuilder: DataDonationAnalyticsWorkBuilder,
+    private val timeCalculation: DataDonationAnalyticsTimeCalculation
+) {
+
+    /**
+     * Enqueue background deadman notification periodic work
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt:23]: deadman?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt [hunk 16]
```diff
@@ -0,0 +1,59 @@
+package de.rki.coronawarnapp.datadonation.analytics.worker
+
+import androidx.work.ExistingPeriodicWorkPolicy
+import androidx.work.ExistingWorkPolicy
+import androidx.work.WorkManager
+import dagger.Reusable
+import javax.inject.Inject
+
+/**
+ * We schedule two workers to allow for randomness in the execution times,
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsScheduler.kt:10]: We just need randomness across the userbase (to be nice to the server), not check on the users device, couldn't we achieve this by setting a random initial delay for the periodic worker?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt [hunk 17]
```diff
@@ -32,15 +33,24 @@ class OnboardingAnalyticsViewModel @AssistedInject constructor(
...
 
     fun onDisableClick() {
-        settings.userInfoAgeGroup.update { PpaData.PPAAgeGroup.AGE_GROUP_UNSPECIFIED }
-        settings.userInfoFederalState.update { PpaData.PPAFederalState.FEDERAL_STATE_UNSPECIFIED }
-        settings.userInfoDistrict.update { 0 }
+        settings.analyticsEnabled.update {
+            false
+        }
+
+        launch {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:48]: If the user smashes that like button, err disable button.... multiple presses can launch this multiple times, right? So donor modules would have to be thread safe on `deleteData`.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @AfterEach
+    fun tearDown() {
+        clearAllMocks()
+    }
+
+    private fun createDonorModules(): Set<DonorModule> = setOf(exposureRiskMetadataDonor)
+
+    private fun createInstance() = spyk(
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:64]: Why do we need  `spy` here?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+        every { settings.analyticsEnabled } returns mockFlowPreference(false)
+        every { analyticsConfig.probabilityToSubmit } returns 0.0
+
+        val analytics = createInstance()
+
+        runBlockingTest2 {
+            analytics.submitIfWanted()
+        }
+
+        // TODO: is there a way to make sure a spy function returned x?
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:101]: You could inject the randomness source and hardcode it to a specific value.

We currently have `SecureRandom` in the DI graph for SafetyNet.

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+            analytics.submitAnalyticsData()
+        }
+    }
+
+    @Test
+    fun `abort due to time since onboarding`() {
+        every { settings.analyticsEnabled } returns mockFlowPreference(false)
+        every { analyticsConfig.probabilityToSubmit } returns 1.0
+        every { settings.lastSubmittedTimestamp } returns mockFlowPreference(
+            Instant.now().minus(Days.TWO.toStandardDuration())
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:133]: Dynamic time in tests is :ghost:

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+
+        val analytics = createInstance()
+
+        runBlockingTest2 {
+            analytics.submitIfWanted()
+        }
+
+        coVerify(exactly = 1) {
+            analytics.submitAnalyticsData()
+            dataDonationAnalyticsServer.uploadAnalyticsData(any())
```
Comment (d4rken) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:188]: Do we have access to the data that would be uploaded such that we can be specific than `any()`?

e.g.

Can we construct `PpaDataRequestAndroid` with the same value? Or would the ProtoBuf equals/hash still not be equal and allow this?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsApiV1.kt [hunk 22]
```diff
@@ -0,0 +1,13 @@
+package de.rki.coronawarnapp.datadonation.analytics.server
+
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaDataRequestAndroid
+import retrofit2.Response
+import retrofit2.http.Body
+import retrofit2.http.POST
+
+interface DataDonationAnalyticsApiV1 {
+    @POST("/version/v1/android/dat")
+    suspend fun submitAndroidAnalytics(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsApiV1.kt:10]: Can I have a test for this stuff too? :cat2: 

Should be straight forward, check the other API tests: Making sure serialization does what it should, bad data leads to an error, and the API path is what we expect.

**Review by d4rken — COMMENTED** [review]
I like what you got

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt (hunk 8 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:104]: we can't roll a 1.0 afaik `nextDouble` is (inclusive, exclusive)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt (hunk 3 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:18]: i talked to @mlenkeit about this, he mentioned that the analytics config part of the app config should not be optional so we can assume a broken config if no analytics config is provided

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt (hunk 4 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:57]: not really, i'll remove the defaults for now as there is no way to instantiate the defaults anyways at the moment

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt (hunk 10 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt:119]: i am relatively sure because of the delay implied in the scheduled background work, but i will add a mutex to be extra safe :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt (hunk 14 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/worker/DataDonationAnalyticsOneTimeWorker.kt:19]: currently the randomness was achieved by scheduling one time work every 24 hours which would then schedule periodic work at a random time in the next 23 hours, but i like the idea of just randomly distributing the initial start delay of the periodic work, will remove the one time worker

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt (hunk 12 above)
Comment (BMItter) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt:23]: I'd prefer @Reuseable

File: Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_datadonation.xml [hunk 23]
```diff
@@ -115,6 +115,98 @@
...
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginTop="@dimen/spacing_tiny"
+                    android:textIsSelectable="true"
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@+id/analytics_last_submit_title"
+                    tools:text="Body" />
+
+            </androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (BMItter) [inline Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_datadonation.xml:147]: please check for layout Stacking (constraint->constraint->...)
I'm not sure here

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+            .create()
+    }
+
+    override suspend fun storeAnalyticsData(analyticsProto: PpaData.PPADataAndroid) =
+        withContext(dispatcherProvider.IO) {
+            if (!analyticsDir.exists()) {
+                analyticsDir.mkdirs()
+            }
+
+            val dataObject = LastAnalyticsSubmission(
```
Comment (BMItter) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/datadonation/analytics/storage/DefaultLastAnalyticsSubmissionLogger.kt:46]: Inject and use TimeStamper 👵

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt (hunk 17 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:48]: hmm true, i am also not 100% happy with the positioning here, as we would need to duplicate the code in the AnalyticsSettings view model later on, maybe also some kind of mutex lockout and we move this to a more generic place where it listens to the analyticsEnabled flow preference?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt (hunk 18 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:64]: i was trying to use a spy to check for the return value of the internal check functions in Analytics i.e. `stopDueToProbabilityToSubmit` to not only check that submission was aborted as it is done now, but make sure that it is aborted for the right reason, maybe throwing a more specific exception in the stop functions and raising the exception handling one level further up in the analytics submission could also help with testability

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt (hunk 19 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:101]: this comment was more in reference to https://github.com/corona-warn-app/cwa-app-android/pull/2280#discussion_r573931976

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt [hunk 25]
```diff
@@ -32,15 +33,24 @@ class OnboardingAnalyticsViewModel @AssistedInject constructor(
     }.asLiveData(dispatcherProvider.IO)
 
     fun onNextButtonClick() {
+        settings.analyticsEnabled.update {
+            true
+        }
+
         completedOnboardingEvent.postValue(Unit)
-        // TODO: ♫ ♪ Hello Kolya, my old friend ♫ ♪
```
Comment (BMItter) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:36]: 👌 😄

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaDataRequestAndroid
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class DataDonationAnalyticsServer @Inject constructor(
+    private val api: Lazy<DataDonationAnalyticsApiV1>
+) {
+
+    suspend fun uploadAnalyticsData(ppaDataRequestAndroid: PpaDataRequestAndroid.PPADataRequestAndroid) {
```
Comment (BMItter) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt:14]: consider to use 'when' instead of various if statements if or when you want (optional)

File: Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_datadonation.xml (hunk 23 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_datadonation.xml:147]: i'm too stupid to do it without as the stacked layout also contains a card border

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt (hunk 26 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/server/DataDonationAnalyticsServer.kt:15]: done :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt (hunk 17 above)
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:48]: Note that FlowPreference is only safe when accessed from the same thread. It's not safe for concurrent access from multiple threads.

I think it makes sense if all deletions, just like starting the analytics submission goes the through `Analytics` main class, and we add a single `Mutex' that guards the deletion and the submission call (as those should not overlap either).

What do you think?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt (hunk 18 above)
Comment (d4rken) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:72]: Maybe set up the `setup()` that the default state is equal to a successful submission, then in each test case (e.g. probability) change a single parameter and check that the process fails appropriately, maybe by checking that no submission happened.

Or if `stopDueToProbabilityToSubmit` causes an abort, then check that `stopDueToNoUserConsent` was called but `stopDueToLastSubmittedTimestamp`, maybe by verifying calls on the mocks used by those functions.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt (hunk 17 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingAnalyticsViewModel.kt:48]: sounds good, i've wrapped changes to the property in calls on `Analytics` which should allow us to execute delete on every settings change

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt (hunk 18 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsTest.kt:72]: I like it :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt (hunk 12 above)
Comment (kolyaopahle) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/AnalyticsModule.kt]: i agree :)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt [hunk 27]
```diff
@@ -5,19 +5,66 @@ import de.rki.coronawarnapp.appconfig.AnalyticsConfig
 import de.rki.coronawarnapp.appconfig.SafetyNetRequirements
 import de.rki.coronawarnapp.appconfig.SafetyNetRequirementsContainer
 import de.rki.coronawarnapp.server.protocols.internal.v2.AppConfigAndroid
+import timber.log.Timber
 import javax.inject.Inject
 
 @Reusable
 class AnalyticsConfigMapper @Inject constructor() : AnalyticsConfig.Mapper {
     override fun map(rawConfig: AppConfigAndroid.ApplicationConfigurationAndroid): AnalyticsConfig {
```
Comment (BMItter) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/appconfig/mapping/AnalyticsConfigMapper.kt:13]: using return if... below would be nice

**Review by BMItr — APPROVED** [review]
one optional comment.
lgtm overall 👌

**Comment by harambasicluka** [comment]
Did on device testing (Pixel 3), logs looked good.

**Review by d4rken — APPROVED** [review]
lgtm
